Mathlib/Analysis/Normed/Algebra/Norm.lean

```diff
@@ -0,0 +1,186 @@
+  the action of `R`.
+* `MulAlgebraNorm` : a multiplicative algebra norm on an `R`-algebra `S` is a multiplicative
+  ring norm on `S` compatible with the action of `R`.
+
+## Tags
+
+norm, algebra norm
+-/
+
+/-- An algebra norm on an `R`-algebra `S` is a ring norm on `S` compatible with the
+action of `R`. -/
+structure AlgebraNorm (R : Type*) [SeminormedCommRing R] (S : Type*) [Ring S] [Algebra R S] extends
+  RingNorm S, Seminorm R S
+
+attribute [nolint docBlame] AlgebraNorm.toSeminorm AlgebraNorm.toRingNorm
+
+instance (K : Type*) [NormedField K] : Inhabited (AlgebraNorm K K) :=
+  ⟨{  toFun     := norm
+      map_zero' := norm_zero
+      add_le'   := norm_add_le
+      neg'      := norm_neg
+      smul'     := norm_mul
+      mul_le'   := norm_mul_le
+      eq_zero_of_map_eq_zero' := fun _ => norm_eq_zero.mp }⟩
+
+/-- `AlgebraNormClass F R S` states that `F` is a type of `R`-algebra norms on the ring `S`.
+You should extend this class when you extend `AlgebraNorm`. -/
+class AlgebraNormClass (F : Type*) (R : outParam <| Type*) [SeminormedCommRing R]
+    (S : outParam <| Type*) [Ring S] [Algebra R S] [FunLike F S ℝ] extends RingNormClass F S ℝ,
+    SeminormClass F R S : Prop
+
+namespace AlgebraNorm
+
+variable {R : Type*} [SeminormedCommRing R] {S : Type*} [Ring S] [Algebra R S] {f : AlgebraNorm R S}
+
+/-- The ring seminorm underlying an algebra norm. -/
+def toRingSeminorm' (f : AlgebraNorm R S) : RingSeminorm S :=
+  f.toRingNorm.toRingSeminorm
+
+instance : FunLike (AlgebraNorm R S) S ℝ where
+  coe f := f.toFun
+  coe_injective' f f' h := by
+    simp only [AddGroupSeminorm.toFun_eq_coe, RingSeminorm.toFun_eq_coe] at h
+    cases f; cases f'; congr;
+    simp only at h
+    ext s
+    erw [h]
+    rfl
+
+instance algebraNormClass : AlgebraNormClass (AlgebraNorm R S) R S where
+  map_zero f        := f.map_zero'
+  map_add_le_add f  := f.add_le'
+  map_mul_le_mul f  := f.mul_le'
+  map_neg_eq_map f  := f.neg'
+  eq_zero_of_map_eq_zero f := f.eq_zero_of_map_eq_zero' _
+  map_smul_eq_mul f := f.smul'
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : CoeFun (AlgebraNorm R S) fun _ => S → ℝ :=
+  DFunLike.hasCoeToFun
+
+theorem toFun_eq_coe (p : AlgebraNorm R S) : p.toFun = p := rfl
+
+@[ext]
+theorem ext {p q : AlgebraNorm R S} : (∀ x, p x = q x) → p = q :=
+  DFunLike.ext p q
+
+/-- An `R`-algebra norm such that `f 1 = 1` extends the norm on `R`. -/
+theorem extends_norm' (hf1 : f 1 = 1) (a : R) : f (a • (1 : S)) = ‖a‖ := by
+  rw [← mul_one ‖a‖, ← hf1]; exact f.smul' _ _
+
+/-- An `R`-algebra norm such that `f 1 = 1` extends the norm on `R`. -/
+theorem extends_norm (hf1 : f 1 = 1) (a : R) : f (algebraMap R S a) = ‖a‖ := by
+  rw [Algebra.algebraMap_eq_smul_one]; exact extends_norm' hf1 _
+
+/-- The restriction of an algebra norm to a subalgebra. -/
+def restriction (A : Subalgebra R S) (f : AlgebraNorm R S) : AlgebraNorm R A where
```

Comment:
Would it make sense to generalize this to any tower of algebras?

Reply:
This is possible if adding `[Nontrivial S]  [NoZeroSMulDivisors A S]` as in the code below, but then I would also have to add hypotheses `[Nontrivial S] [NoZeroDivisors S]` to `eq_of_powMul_faithful`. Do you think this is worth it?

```lean
/-- The restriction of an algebra norm in a scalar tower. -/
def AlgebraNorm.restriction {A : Type*} [CommRing A] [Algebra R A] [Algebra A S]
  [Nontrivial S]  [NoZeroSMulDivisors A S] [IsScalarTower R A S] (f : AlgebraNorm R S) : 
    AlgebraNorm R A where
  toFun       := fun x : A => f (algebraMap A S x)
  map_zero'   := by simp only [map_zero]
  add_le' x y := by simp only [map_add, map_add_le_add]
  neg' x      := by simp only [map_neg, map_neg_eq_map]
  mul_le' x y := by simp only [map_mul, map_mul_le_mul]
  eq_zero_of_map_eq_zero' x hx := by
    rw [← map_eq_zero_iff (algebraMap A S) (NoZeroSMulDivisors.algebraMap_injective A S)]
    exact eq_zero_of_map_eq_zero f hx
  smul' r x := by
    simp only [Algebra.smul_def, map_mul, ← IsScalarTower.algebraMap_apply]
    simp only [← smul_eq_mul, algebraMap_smul, map_smul_eq_mul]
```

Reply:
Hmm, probably that isn't worth it. I could imagine it is useful to have a version for algebras where you explicitly ask for the assumption `Function.Injective (algebraMap A S)`. Feel free to add that if you think it's useful. Otherwise, let's just leave it like this.

Reply:
I have added it as `AlgebraNorm.isScalarTower_restriction`.